OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,1142 @@
+public:
+  using size_t                     = std::size_t;
+  using string                     = std::string;
+  using MetaDataType               = std::unordered_map<string, 
+                                                        SimTK::AbstractValue*>;
+  using ColLabelsType              = std::unordered_map<string, size_t>;
+  using ColLabelsIterType          = ColLabelsType::iterator;
+  using ColLabelsConstIterType     = ColLabelsType::const_iterator;
+  using ColLabelsIterPairType      = std::pair<ColLabelsIterType,
+                                               ColLabelsIterType>;
+  using ColLabelsConstIterPairType = std::pair<ColLabelsConstIterType,
+                                               ColLabelsConstIterType>;
+
+  /// Construct empty DataTable.
+  DataTable_() :
+    data{}, metadata{}, col_ind{} {}
+
+  /// Construct DataTable of given shape populated with given value val. 
+  /// Default value for val is NaN.
+  /// \param nrows Number of rows.
+  /// \param ncols Number of columns.
+  /// \param val   Value to initialize all the entries to.
+  DataTable_(size_t nrows,
+             size_t ncols,
+             const ET& val = ET{SimTK::NaN}) :
+    data{int(nrows), int(ncols), val}, metadata{}, col_ind{} {}
+
+  /// Construct DataTable using an iterator(satisfying requirement of an 
+  /// input_iterator) which produces one entry at a time. The entries of 
+  /// DataTable_ are copy initialized using the values produced by the 
+  /// iterator. For example, specifying ROWWISE for parameter dir and 10 for 
+  /// parameter ndir will populate the DataTable one row at a time with each 
+  /// row's length taken to be 10 elements.
+  ///
+  /// \param first Beginning of range covered by the iterator.
+  /// \param last End of the range covered by the iterator.
+  /// \param ndir Extent of the dimention specified by parameter dir. 
+  /// \param dir Dimension to populate the DataTable. Populate row-wise or 
+  ///              column wise. See OpenSim::InputItDir for possible values.
+  /// \param allow_missing Allow for missing values. When set to false, this
+  ///                      function will throw if the input iterator fills up 
+  ///                      the last row/col partially. When set to true, 
+  ///                      missing elements are set to SimTK::NaN.
+  ///
+  /// \throws ZeroElements When input-iterator does not produce any elements.
+  ///                      That is first == last.
+  template<typename InputIt>
+  DataTable_(InputIt first,
+             typename std::enable_if<!std::is_integral<InputIt>::value,
+                                     InputIt>::type last,
+             const size_t ndir,
+             const InputItDir dir = ROWWISE,
+             const bool allow_missing = false) :
+    data{static_cast<int>(dir == ROWWISE ? 1    : ndir), 
+         static_cast<int>(dir == ROWWISE ? ndir :    1)},
+    metadata{},
+    col_ind{} {
+    if(!(first != last))
```

Comment:
The [documentation I can find](http://en.cppreference.com/w/cpp/concept/InputIterator) says an InputIterator has to be EqualityComparable, meaning it should support `operator==()`. That is also consistent with the fact that `<utility>` contains generic templatized definitions for operators `!= <= >= >` defined in terms of `==` and `<` which are what C++ normally requires. So I think you might have this reversed.

I was able to find a draft standard and it agrees:
```
24.2.3 Input iterators [input.iterators]
1 A class or a built-in type X satisfies the requirements of an input iterator for the value type T if X satisfies
the Iterator (24.2.2) and EqualityComparable (Table 17) requirements and the expressions in Table 107
are valid and have the indicated semantics.
```
All mentions there were of `==` rather than `!=`.